File: Data Structures & Algorithms/add-two-numbers/submission-1.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var overflow = 0
        val result = ListNode(0)
        var current: ListNode? = result

        var first = l1
        var second = l2

        while(first != null || second != null) {
            val l1val = first?.`val` ?: 0
            val l2val = second?.`val` ?: 0
            val currentTotal = l1val + l2val + overflow
            
            val withoutOverflow = currentTotal % 10
            val newOverflow = currentTotal / 10

            println("$withoutOverflow | $newOverflow")

            current?.next = ListNode(withoutOverflow)
            current = current?.next

            overflow = newOverflow

            first = first?.next
            second = second?.next
        }

        if(overflow != 0) {
            current?.next = ListNode(overflow)
        }

        return result.next
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        res = lists
        while len(res) > 1:
            list1, list2 = res.pop(), res.pop()
            result = self.merge_lists(list1, list2)
            res.append(result)

        return res[0]
        
    def merge_lists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        result = dummy = ListNode()

        while list1 and list2:
            if list1.val < list2.val:
                dummy.next = list1
                list1 = list1.next
            else:
                dummy.next = list2
                list2 = list2.next
            dummy = dummy.next

        dummy.next = list1 or list2

        return result.next